### ISSUE TITLE
Creating a Unix Domain Socket properly fails when done with ScalaNative

### ISSUE THREAD

**Issue by markehammons** [scala-native/scala-native#2899]
My scala code looked like: 
```scala
        listenSocket <- socket
            .socket(socket.AF_UNIX, socket.SOCK_DGRAM, 0)
            .cCall("create listen socket")
         _ <- socket.bind(listenSocket, addr, 16.toUInt).cCall("bind socket")
```
In the non-working version, I pass the correct size (110) to bind, and I would get the error `Left(bind socket: 97 - Address family not supported by protocol)
[3:15 PM]
`
The same code written in C doesn't display this problem at all:

```c
int getUDS(char * name) {
  struct sockaddr_un addr;
  addr.sun_family = AF_UNIX;
  strncpy(addr.sun_path, name, sizeof(addr.sun_path) - 1);

  int socketResult = socket(AF_UNIX, SOCK_DGRAM, 0);
  if(socketResult == -1) {
    int errnum = errno;
    perror(strerror(errnum));
    return -1;
  }

  int bindResult = bind(socketResult, (struct sockaddr*) &addr, sizeof(struct sockaddr_un));
  if(bindResult == -1) {
    int errnum = errno;
    perror(strerror(errnum));
    return -1;
  }

  return socketResult;
}
```

**Comment by armanbilge** [comment]
Yes, unfortunately this is broken in 0.4.x but I believe should be fixed in 0.5.x. I worked around it by writing my own binding for `bind`.

https://github.com/armanbilge/fs2-io_uring/blob/2d7df694371cb0fa0fdd33e8966b72274106f8d7/uring/src/main/scala/fs2/io/uring/unsafe/syssocket.scala#L24-L25

**Comment by ekrich** [comment]
This can't be fixed for `0.4.x`?

**Comment by armanbilge** [comment]
I appreciated the detailed explanation. It documents new techniques for glue code that I hope we will continue to adopt going forward.

**Comment by LeeTibbert** [comment]
@ekrich , I apologize profusely if I caused any offense. That was not intended in any way.

@markehammons Thank you for reporting this problem, along with a reproducer.  I am
sorry that you experienced it and that a backport is not readily available.

**Comment by LeeTibbert** [comment]
@markehammons  On looking at the Scala code more closely & considering issues of 'direct passthru" how is
storage for `addr` declared?  Have you defined your own structure or Array[Byte]?  Is it large enough?

The Open Group (POSIX) has a sockaddr_un specified in `<sys/un.h>`. A quick search of SN
posixlib seems to indicate that "sys/un.scala" is not implemented. 

 At first blush, that looks like it could be implemented in 0.5.0 and backported to 0.4.x.
 That would not help you today or tomorrow, but some good might come from your detecting this "feature". 
Implementing "un.h"  would not solve the "bind()" issue, but it would put compile-time guards on the sockaddr_un
declaration.  That size is quite likely to vary by PS (and not exist on Windows).

If you think "un.h" would be helpful in your medium range work (i.e. if you will be working with unix sockets),
I can create an Issue so that the idea stays "somewhere near the top of the heap".

**Comment by armanbilge** [comment]
If it's any help, I also have `un.h` here:
https://github.com/armanbilge/fs2-io_uring/blob/2d7df694371cb0fa0fdd33e8966b72274106f8d7/uring/src/main/scala/fs2/io/uring/unsafe/sysun.scala

**Comment by markehammons** [comment]
@LeeTibbert I tried two solutions. I had the struct type declared in a un.scala directly for Linux, and I had more general code where a c function took the struct data, and initialized the struct, casting it to sockaddr when handing it back to Scala.

sockaddr_un highlights a problem that Scala native will brush against more and more, platform dependent types and platform dependent structs. The size of sockaddr_un can vary a lot. It's 90 bytes on some platforms supposedly. sockaddr_un could have a well made scala definition via transparent inline macros, but only because the first parameter doesn't ever change in size.

**Comment by LeeTibbert** [comment]
@markehammons  

Your description of the solutions you have tried & what works is good info to have. Thank you.

Your second paragraph is spot on!  

Let me hang my head in, a little bit of, shame before I describe what I tend to do.
I am well aware that others may find this approach (beyond) horrid.

For better or worse, I have always taken the "largest size on defined supported systems" approach.
That is, pick a size I know works  on Linux & macOS. I have "guard" code testing that the
size of the Scala Native structure is equal or larger then the compile size on any "known" system.
It also reports "unknown" systems.   The latter goes against "do not preclude unknown systems" but also
keeps memory spray at bay on unknown systems.  A balancing act.

This balances for implementation correctness and speed, rather than "minimal memory use"
(I can hear David B saying "Yeah, that is obvious, you take my 20G for a simple regex")

Because of the rate of change in the basic Scala implementation, especially macros, I am reluctant
to spend too much developer time on point optimizations. (yeah, they say, it shows in the lack-of runtime performance
of your work).   I figure it is much easier for a motivated & talented person to optimize a small, well defined piece of
code once it exists (and had been shown by benchmarks to be sub-optimal).

**Comment by LeeTibbert** [comment]
@markehammons @armanbilge 

You might find this interesting.

Later in the day I was ratting in `socket.scala` for some javalib changes I am
studying.  Turns out the 'maximum size' issue of this topic is moot.

POSIX declares a `sockaddr_storage` which is defined as holding 
as much, or more than, enough storage for the largest sockaddr
on a given system.

That code is, honor to my predecessors, just plain broken in
0.4.x.  However, an elf fixed it for 0.5.0.  

In 0.5.x, one ought to be able to stackalloc a sockaddr_storage
and cast the pointer to what you need.

I am always reluctant to define 'easy' before having done
a task, but I think it would be relatively easy for me to 
port that specific change back to 0.4.x, if that would 
make your life easier.  Does not solve today or this
years problem but could brighten the future. 


sockaddr_storage is 120 (or so) bytes, so, IMO, it is beginning
to verge on needing to be allocated with `alloc` (heap) rather
than `stackalloc` (stack).

**Comment by markehammons** [comment]
> @markehammons @armanbilge
> 
> You might find this interesting.
> 
> Later in the day I was ratting in `socket.scala` for some javalib changes I am studying. Turns out the 'maximum size' issue of this topic is moot.
> 
> POSIX declares a `sockaddr_storage` which is defined as holding as much, or more than, enough storage for the largest sockaddr on a given system.
> 
> That code is, honor to my predecessors, just plain broken in 0.4.x. However, an elf fixed it for 0.5.0.
> 
> In 0.5.x, one ought to be able to stackalloc a sockaddr_storage and cast the pointer to what you need.
> 
> I am always reluctant to define 'easy' before having done a task, but I think it would be relatively easy for me to port that specific change back to 0.4.x, if that would make your life easier. Does not solve today or this years problem but could brighten the future.
> 
> sockaddr_storage is 120 (or so) bytes, so, IMO, it is beginning to verge on needing to be allocated with `alloc` (heap) rather than `stackalloc` (stack).
> 
> I hope your work is going well and that this defect has not derailed you for too long.

I already worked around it by doing the binding code in C. This hasn't really proven to be a problem elsewhere, and I'm chugging along fine without the fix. The only derailment that occurred was the day I spent thinking this was a bug in my code and not scala native :laughing:

**Comment by LeeTibbert** [comment]
@markehammons  Motivated by this Issue and its helpful discussions, I created PR #3025.
I believe that will, once merged, close this issue.   

While I can not give you back the day you ~wasted~ invested, you/we have made it less likely
that others will fall into the same trap.   

I do hope you continue your exploration of Scala Native. I standby to help, if possible,
so that your time is learning, not frustration.  It is only by people exploring the the
unfrequented parts of Scala Native that it gets better.

TL;DR;   I supplied a `un.scala` loosely based on Arman's work (Thank you, Arman)
              and a corresponding `un.c` file with `_Static_assert` statements to assure
              the proper sizes on Linux, macOS, & Windows.
